## **ASSOCIATION RULES**

In [18]:
#installing apyori library
!pip install apyori

Defaulting to user installation because normal site-packages is not writeable


In [19]:
#Loading dataset
import pandas as pd
df = pd.read_csv("Online retail.csv",header=None)
df

0                  1            2                 3   \
0            shrimp            almonds      avocado    vegetables mix   
1           burgers          meatballs         eggs               NaN   
2           chutney                NaN          NaN               NaN   
3            turkey            avocado          NaN               NaN   
4     mineral water               milk   energy bar  whole wheat rice   
...             ...                ...          ...               ...   
7496         butter         light mayo  fresh bread               NaN   
7497        burgers  frozen vegetables         eggs      french fries   
7498        chicken                NaN          NaN               NaN   
7499       escalope          green tea          NaN               NaN   
7500           eggs    frozen smoothie  yogurt cake    low fat yogurt   

                4                 5     6               7             8   \
0     green grapes  whole weat flour  yams  cottage cheese  energy drink   
1              NaN               NaN   NaN             NaN           NaN   
2              NaN               NaN   NaN             NaN           NaN   
3              NaN               NaN   NaN             NaN           NaN   
4        green tea               NaN   NaN             NaN           NaN   
...            ...               ...   ...             ...           ...   
7496           NaN               NaN   NaN             NaN           NaN   
7497     magazines         green tea   NaN             NaN           NaN   
7498           NaN               NaN   NaN             NaN           NaN   
7499           NaN               NaN   NaN             NaN           NaN   
7500           NaN               NaN   NaN             NaN           NaN   

                9               10         11     12     13             14  \
0     tomato juice  low fat yogurt  green tea  honey  salad  mineral water   
1              NaN             NaN        NaN    NaN    NaN            NaN   
2              NaN             NaN        NaN    NaN    NaN            NaN   
3              NaN             NaN        NaN    NaN    NaN            NaN   
4              NaN             NaN        NaN    NaN    NaN            NaN   
...            ...             ...        ...    ...    ...            ...   
7496           NaN             NaN        NaN    NaN    NaN            NaN   
7497           NaN             NaN        NaN    NaN    NaN            NaN   
7498           NaN             NaN        NaN    NaN    NaN            NaN   
7499           NaN             NaN        NaN    NaN    NaN            NaN   
7500           NaN             NaN        NaN    NaN    NaN            NaN   

          15                 16               17       18         19  
0     salmon  antioxydant juice  frozen smoothie  spinach  olive oil  
1        NaN                NaN              NaN      NaN        NaN  
2        NaN                NaN              NaN      NaN        NaN  
3        NaN                NaN              NaN      NaN        NaN  
4        NaN                NaN              NaN      NaN        NaN  
...      ...                ...              ...      ...        ...  
7496     NaN                NaN              NaN      NaN        NaN  
7497     NaN                NaN              NaN      NaN        NaN  
7498     NaN                NaN              NaN      NaN        NaN  
7499     NaN                NaN              NaN      NaN        NaN  
7500     NaN                NaN              NaN      NaN        NaN  

[7501 rows x 20 columns]

### EDA

In [20]:
df.shape

(7501, 20)

In [21]:
df.tail()

0                  1            2               3          4   \
7496    butter         light mayo  fresh bread             NaN        NaN   
7497   burgers  frozen vegetables         eggs    french fries  magazines   
7498   chicken                NaN          NaN             NaN        NaN   
7499  escalope          green tea          NaN             NaN        NaN   
7500      eggs    frozen smoothie  yogurt cake  low fat yogurt        NaN   

             5    6    7    8    9    10   11   12   13   14   15   16   17  \
7496        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
7497  green tea  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
7498        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
7499        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
7500        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

       18   19  
7496  NaN  NaN  
7497  NaN  NaN  
7498  NaN  NaN  
7499  NaN  NaN  
7500  NaN  NaN

In [22]:
# check for duplicates
duplicates = df[df.duplicated()]
print("Duplicate Rows:")
duplicates

Duplicate Rows:


0           1    2    3    4    5    6    7    8    9    10  \
34          cookies         NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
42        spaghetti         NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
60        spaghetti         NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
64           turkey        eggs  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
65     french fries         NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
...             ...         ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
7491  herb & pepper         NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
7492      chocolate    escalope  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
7495       pancakes  light mayo  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
7498        chicken         NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
7499       escalope   green tea  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

       11   12   13   14   15   16   17   18   19  
34    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
42    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
60    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
64    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
65    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
...   ...  ...  ...  ...  ...  ...  ...  ...  ...  
7491  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
7492  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
7495  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
7498  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
7499  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[2325 rows x 20 columns]

In [23]:
# remove duplicate rows

df = df.drop_duplicates()
df

0                  1             2                 3   \
0            shrimp            almonds       avocado    vegetables mix   
1           burgers          meatballs          eggs               NaN   
2           chutney                NaN           NaN               NaN   
3            turkey            avocado           NaN               NaN   
4     mineral water               milk    energy bar  whole wheat rice   
...             ...                ...           ...               ...   
7493        burgers             salmon      pancakes      french fries   
7494         turkey            burgers  dessert wine            shrimp   
7496         butter         light mayo   fresh bread               NaN   
7497        burgers  frozen vegetables          eggs      french fries   
7500           eggs    frozen smoothie   yogurt cake    low fat yogurt   

                   4                 5       6               7             8   \
0        green grapes  whole weat flour    yams  cottage cheese  energy drink   
1                 NaN               NaN     NaN             NaN           NaN   
2                 NaN               NaN     NaN             NaN           NaN   
3                 NaN               NaN     NaN             NaN           NaN   
4           green tea               NaN     NaN             NaN           NaN   
...               ...               ...     ...             ...           ...   
7493  frozen smoothie       fresh bread    mint             NaN           NaN   
7494            pasta          tomatoes  pepper            milk      pancakes   
7496              NaN               NaN     NaN             NaN           NaN   
7497        magazines         green tea     NaN             NaN           NaN   
7500              NaN               NaN     NaN             NaN           NaN   

                    9               10               11     12     13  \
0         tomato juice  low fat yogurt        green tea  honey  salad   
1                  NaN             NaN              NaN    NaN    NaN   
2                  NaN             NaN              NaN    NaN    NaN   
3                  NaN             NaN              NaN    NaN    NaN   
4                  NaN             NaN              NaN    NaN    NaN   
...                ...             ...              ...    ...    ...   
7493               NaN             NaN              NaN    NaN    NaN   
7494  whole wheat rice             oil  frozen smoothie    NaN    NaN   
7496               NaN             NaN              NaN    NaN    NaN   
7497               NaN             NaN              NaN    NaN    NaN   
7500               NaN             NaN              NaN    NaN    NaN   

                 14      15                 16               17       18  \
0     mineral water  salmon  antioxydant juice  frozen smoothie  spinach   
1               NaN     NaN                NaN              NaN      NaN   
2               NaN     NaN                NaN              NaN      NaN   
3               NaN     NaN                NaN              NaN      NaN   
4               NaN     NaN                NaN              NaN      NaN   
...             ...     ...                ...              ...      ...   
7493            NaN     NaN                NaN              NaN      NaN   
7494            NaN     NaN                NaN              NaN      NaN   
7496            NaN     NaN                NaN              NaN      NaN   
7497            NaN     NaN                NaN              NaN      NaN   
7500            NaN     NaN                NaN              NaN      NaN   

             19  
0     olive oil  
1           NaN  
2           NaN  
3           NaN  
4           NaN  
...         ...  
7493        NaN  
7494        NaN  
7496        NaN  
7497        NaN  
7500        NaN  

[5176 rows x 20 columns]

In [24]:
#check for duplicates
duplicates = df[df.duplicated()]
print("Duplicate Rows:")
duplicates

Duplicate Rows:


Empty DataFrame
Columns: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19]
Index: []

In [25]:
df.shape

(5176, 20)

In [26]:
# Check for rows with all null values
all_null_rows = df.isnull().all(axis=1)
if all_null_rows.any():
    print("DataFrame contains rows with all null values.")
else:
    print("DataFrame does not contain rows with all null values.")

DataFrame does not contain rows with all null values.


In [27]:
#getting transaction list
trans = []
for i in range(0, 5176):
  trans.append([str(df.values[i,j]) for j in range(0, 20)])
trans



[['shrimp',
  'almonds',
  'avocado',
  'vegetables mix',
  'green grapes',
  'whole weat flour',
  'yams',
  'cottage cheese',
  'energy drink',
  'tomato juice',
  'low fat yogurt',
  'green tea',
  'honey',
  'salad',
  'mineral water',
  'salmon',
  'antioxydant juice',
  'frozen smoothie',
  'spinach',
  'olive oil'],
 ['burgers',
  'meatballs',
  'eggs',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan'],
 ['chutney',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan'],
 ['turkey',
  'avocado',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan',
  'nan'],
 ['mineral water',
  'milk',
  'energy bar',
  'whole wheat rice',
  'green tea',
  'nan',
  'nan',
  'nan',
 

In [28]:
len(trans)

5176

In [29]:

# getting the assocation rule report with appropriate threshold
from apyori import apriori
results = apriori(transactions=trans,min_support = 0.003,
                               min_confidence = 0.2,
                               min_lift = 2,
                               min_length = 2,
                               max_length = 2)

report = list(results)
report


[RelationRecord(items=frozenset({'almonds', 'burgers'}), support=0.007534775888717156, ordered_statistics=[OrderedStatistic(items_base=frozenset({'almonds'}), items_add=frozenset({'burgers'}), confidence=0.2565789473684211, lift=2.2547582879099277)]),
 RelationRecord(items=frozenset({'turkey', 'barbecue sauce'}), support=0.0036707882534775887, ordered_statistics=[OrderedStatistic(items_base=frozenset({'barbecue sauce'}), items_add=frozenset({'turkey'}), confidence=0.2375, lift=2.852204176334107)]),
 RelationRecord(items=frozenset({'extra dark chocolate', 'chicken'}), support=0.004057187017001545, ordered_statistics=[OrderedStatistic(items_base=frozenset({'extra dark chocolate'}), items_add=frozenset({'chicken'}), confidence=0.23863636363636362, lift=2.8460410557184748)]),
 RelationRecord(items=frozenset({'chicken', 'light cream'}), support=0.006568778979907265, ordered_statistics=[OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.293

In [30]:
#length of report
len(report)

18

In [31]:

#print individual report value
print(report[0])
print('support:',report[0][1]) # support
print('base item:',report[0][2][0][0]) # base
print('add item:',report[0][2][0][1]) # add
print('confidence:',report[0][2][0][2]) # confidence
print('lift:',report[0][2][0][3]) # lift



RelationRecord(items=frozenset({'almonds', 'burgers'}), support=0.007534775888717156, ordered_statistics=[OrderedStatistic(items_base=frozenset({'almonds'}), items_add=frozenset({'burgers'}), confidence=0.2565789473684211, lift=2.2547582879099277)])
support: 0.007534775888717156
base item: frozenset({'almonds'})
add item: frozenset({'burgers'})
confidence: 0.2565789473684211
lift: 2.2547582879099277


In [32]:
#getting reports and storing in list

s = []
b = []
a = []
c = []
l = []

for i in range(0,18):
    s.append(report[i][1]) # suport
    b.append(report[i][2][0][0]) # base
    a.append(report[i][2][0][1]) # add
    c.append(report[i][2][0][2]) # confidence
    l.append(report[i][2][0][3]) # lift



In [33]:
#converting lists into dataframe
S_pd = pd.DataFrame(s)
B_pd = pd.DataFrame(b)
A_pd = pd.DataFrame(a)
C_pd = pd.DataFrame(c)
L_pd = pd.DataFrame(l)




In [34]:
#concat the dataframe and finalize the association report
data = pd.concat([B_pd,A_pd,S_pd,C_pd,L_pd],axis=1)
data.columns = ["Base","Add",'Support','Confidence','Lift']

data

Base                Add   Support  Confidence      Lift
0                almonds            burgers  0.007535    0.256579  2.254758
1         barbecue sauce             turkey  0.003671    0.237500  2.852204
2   extra dark chocolate            chicken  0.004057    0.238636  2.846041
3            light cream            chicken  0.006569    0.293103  3.495630
4   mushroom cream sauce           escalope  0.006955    0.288000  3.434765
5                  pasta           escalope  0.005023    0.288889  3.445366
6   extra dark chocolate          olive oil  0.003864    0.227273  2.591109
7          fromage blanc              honey  0.004830    0.247525  3.790497
8        parmesan cheese  frozen vegetables  0.007921    0.284722  2.193039
9        gluten free bar           pancakes  0.003091    0.333333  2.662551
10         herb & pepper        ground beef  0.022798    0.343023  2.525588
11          tomato sauce        ground beef  0.007728    0.384615  2.831820
12           light cream          olive oil  0.004637    0.206897  2.358803
13     whole wheat pasta               milk  0.013910    0.342857  2.016623
14          tomato sauce          olive oil  0.004057    0.201923  2.302101
15     whole wheat pasta          olive oil  0.011012    0.271429  3.094525
16                 pasta             shrimp  0.007148    0.411111  4.156076
17         strong cheese          spaghetti  0.005410    0.482759  2.103332

#### **INSIGHTS**

**Support:** The proportion of transactions in which the base and add items co-occur. Higher support indicates more frequent occurrence.

**Confidence:** The probability of purchasing the "add" item given the "base" item. Higher confidence suggests stronger associations.

**Lift:** A measure of how much more likely the "add" item is purchased when the "base" item is already purchased compared to random chance. A lift > 1 indicates a positive association.

**1. High Lift Values Suggest Strong Associations**

1) Fromage blanc → Honey:

**Lift** = 3.79: Customers who buy "fromage blanc" are 3.79 times more likely to purchase "honey" compared to random chance.
**Support** = 0.0048: While not a frequent combination, this pair indicates a strong preference among a niche group of customers.
**Insight:** Market these items together, especially in gourmet or health food sections.

2) Pasta → Shrimp:

**Lift** = 4.16: Customers buying "pasta" are 4.16 times more likely to also buy "shrimp."
**Confidence** = 0.411: There is a 41% chance of purchasing shrimp when pasta is purchased.
**Insight:** Bundle pasta and shrimp as a combo offer for meal preparation, like a seafood pasta kit.

**2. Frequent Associations with Moderate Lift**

1) Herb & Pepper → Ground Beef:

**Support** = 0.0228: This is one of the most frequent pairs.
**Lift** = 2.53: Buying "herb & pepper" makes it 2.53 times more likely that "ground beef" is also purchased.
**Insight:** These items could be marketed together in recipe suggestions or displayed nearby in-store.

2) Tomato Sauce → Ground Beef:

**Lift** = 2.83: A strong association suggests these items are commonly used together in recipes like pasta or lasagna.
**Insight:** Highlight recipes involving these ingredients, such as spaghetti bolognese, to increase sales.

**3. Niche and Specialized Purchases**

1) Extra Dark Chocolate → Olive Oil:

**Lift** = 2.59: A positive association, possibly linked to health-conscious customers or gourmet cooking preferences.
**Insight:** These items could be cross-promoted in a health or luxury section.

2) Gluten-Free Bar → Pancakes:

**Lift** = 2.66: Suggests customers buying gluten-free bars are interested in alternative breakfast options.
**Insight:** Create a gluten-free breakfast section that includes both bars and pancake mix.

**4. Health-Conscious and Gourmet Trends**

1) Light Cream → Chicken:

**Lift** = 3.49: Indicates a strong association, possibly for creamy chicken recipes.
**Insight:** Promote recipes or bundles featuring light cream and chicken.

2) Whole Wheat Pasta → Olive Oil:

**Lift** = 3.09: Indicates customers pairing healthy pasta alternatives with olive oil.
**Insight:** Highlight these items in health food aisles or bundle them with Mediterranean diet suggestions.

**5. Low Lift and High Confidence**

1) Strong Cheese → Spaghetti:

**Confidence** = 48.27%: Almost half of the transactions with "strong cheese" also include "spaghetti."
**Lift** = 2.10: The association is positive but not exceptionally strong.
**Insight:** This pairing aligns with Italian cuisine preferences; consider marketing them for pasta lovers.

**1.	What is lift and why is it important in Association rules?**

Lift is a metric used to evaluate the strength of an association rule in data mining. It measures how much more likely two (or more) items are to occur together than they would if they were statistically independent.

Formula for Lift:

For a rule A→B, the lift is calculated as:

**Lift(A→B)=Support(A∩B)/Support(A)×Support(B)**

**--> Why Is Lift Important?**

Lift is crucial because it provides insights into the actual significance of an association rule, beyond mere co-occurrence.

1)Indicates True Association:

2)Normalized Measure:

3)Useful for Ranking Rules:



**2.	What is support and Confidence. How do you calculate them?**

--> What is Support?

Support measures the frequency of an itemset occurring in the dataset. It helps identify how often an item or a combination of items appears in transactions.

Formula for Support:

For an itemset A :

**Support(A)=Number of transactions containing A/Total number of transactions**

--> What is Confidence?

Confidence measures the likelihood of occurrence of an item B, given that item A has already occurred. It quantifies the predictive power of an association rule.

Formula for Confidence:

For a rule A→B:

**Confidence(A→B)=Support(A∩B)/Support(A)**


**3.	What are some limitations or challenges of Association rules mining?**

Association rule mining is a powerful technique for discovering relationships between items in a dataset. However, it comes with several limitations and challenges that can affect its applicability and effectiveness. Here are the main ones:
________________________________________
1. Scalability and Performance Issues
•	Large Datasets: As the dataset grows in size (number of transactions or items), the computational complexity of generating frequent itemsets and rules increases significantly.
•	High Dimensionality: Datasets with many items can lead to a combinatorial explosion in the number of possible itemsets, making the process resource-intensive.
________________________________________
2. Handling Sparse Data
•	In datasets where most transactions contain only a few items (e.g., online shopping carts), the frequent itemsets might be too few, and mining meaningful associations can be challenging.
________________________________________
3. Setting Thresholds (Support and Confidence)
•	Support Threshold: If set too high, interesting but less frequent patterns may be missed. If set too low, it may generate too many itemsets, including irrelevant ones.
•	Confidence Threshold: A high confidence threshold might filter out potentially valuable rules, while a low threshold might include unreliable associations.
________________________________________
4. Lack of Contextual Understanding
•	Association rules do not consider the context or causality of the relationships. For instance, a rule A→BA \rightarrow BA→B doesn't imply that AAA causes BBB; it only indicates co-occurrence.
•	Rules may sometimes reflect spurious or coincidental relationships.
________________________________________
5. Difficulty with Continuous or Quantitative Data
•	Association rule mining works best with categorical data. For continuous or numerical attributes, data must be discretized (e.g., converting age into age ranges), which can lead to a loss of information or introduce bias.
________________________________________
6. Handling Null or Missing Data
•	Missing values in datasets can interfere with the identification of frequent patterns. Decisions about whether to exclude, fill, or ignore missing values can impact results.



